python/taichi/lang/expr.py

```diff
@@ -17,16 +18,21 @@ def __init__(self, *args, tb=None):
                 self.ptr = args[0].ptr
                 self.tb = args[0].tb
             elif is_taichi_class(args[0]):
-                raise ValueError('cannot initialize scalar expression from '
-                                 f'taichi class: {type(args[0])}')
+                raise TaichiTypeError(
+                    'Cannot initialize scalar expression from '
+                    f'taichi class: {type(args[0])}')
             else:
                 # assume to be constant
                 arg = args[0]
-                try:
-                    if isinstance(arg, np.ndarray):
-                        arg = arg.dtype(arg)
-                except:
-                    pass
+                if isinstance(arg, np.ndarray):
+                    if arg.size != 1:
+                        raise TaichiTypeError(
+                            "Only numpy array with one element can be used to initialize a scalar expression"
+                        )
+                    try:
+                        arg = arg.dtype.type(arg.reshape(()))
```

Comment:
Hmm I'm a little skeptical about this: are we going to allow arbitrary shape with one element to be converted to constant expr? 

IMHO we implicitly lose the shape information by doing this without user awareness. Is it better to let users do `arg.reshape(())` instead of us doing it?

Reply:
There is a case in our tests (tests/python/test_loops.py::test_numpy_loops), should we ban it and fix the test?

Reply:
Discussed offline: let's make it stricter for now, users should be aware of the numpy shape and make sure it only pass in a scalar :D